javascript/implementation/KeySet.ts

```diff
@@ -0,0 +1,228 @@
+        }
+    }
+
+    /**
+     * Adds a key to the keyset.
+     * If a bundler is supplied, the function will add the entry to that bundler
+     * and return immediately (presumably you know what to do with a CS if you passed it in).
+     * If the caller does not supply a bundler, then one is created on the fly, and
+     * then this method will await on the CS being added to the database instance.
+     * This is to allow simple console usage like:
+     *      await myKeySet.add("foo");
+     * @param key
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to the address of the newly created entry
+     */
+    async add(key: KeyType, change?: Bundler|string): Promise<Muid> {
+        return await this.addEntry(key, Container.INCLUSION, change);
+    }
+
+    /**
+     * Similar to add method, but for multiple entries.
+     * @param keys an iterable of keys to add to the key set
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to a set of Muids for the created entries.
+     */
+    async update(keys: Iterable<KeyType>, change?: Bundler|string): Promise<Set<Muid>> {
+        let bundler: Bundler;
+        if (change instanceof Bundler) {
+            bundler = change;
+        } else {
+            bundler = new Bundler(change);
+        }
+        const additions = new Set<Muid>;
+        for (const key of keys) {
+            additions.add(await this.addEntry(key, Container.INCLUSION, bundler));
+        }
+        await this.ginkInstance.addBundler(bundler);
+        return additions;
+    }
+
+    /**
+     * Tries to delete the key and return the key itself, throws an error if key not found.
+     * @param key one key of KeyType
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to a key.
+     */
+    async pop(key: KeyType, change?: Bundler|string): Promise<KeyType> {
+        if (await this.has(key)) {
+            await this.delete(key, change)
+            return key;
+        } else {
+            throw new Error("Key not found");
+        }
+    }
+
+    /**
+     * Adds a deletion marker (tombstone) for a particular key in the directory.
+     * The corresponding value will be seen to be unset in the data model.
+     * @param key
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to the address of the newly created deletion entry
+     */
+    async delete(key: KeyType, change?: Bundler|string): Promise<Muid> {
+        return await this.addEntry(key, Container.DELETION, change);
+    }
+
+    /**
+     * Function to iterate over the contents of the key set, showing the address of each entry (which can be used in pop).
+     * @param asOf
+     * @returns an async iterator across everything in the key set, with values returned as pairs of Muid, Key
+     */
+    entries(asOf?: AsOf): AsyncGenerator<[Muid,KeyType], void, unknown> {
```

Comment:
This needs to match the interface of the built in javascript Set (other than being async), which means that it should return pairs [key, key].  There shouldn't be a way to remove keys via MUID, because unlike a sequence where you can have one value in the collection twice, in a set it can be at most once.